### PR TITLE
Add `AppEnv` component and some small improvements

### PR DESCRIPTION
**Pull request by sweh**

### CODE REVIEW

**Comment by ctheune** [comment]
Hey @wosc, what do you think about this?

File: CHANGES.md [hunk 1]
```diff
@@ -6,6 +6,14 @@
 
 - Update embedded `appenv` to support Python 3.4+.
 
+- Add component `AppEnv` to manage virtualenvs similar to the `appenv` package
+  superseding previous virtualenv and buildout components.
+
+- Allow using `assert` statements istead if `batou.UpdateNeeded`.
```
Comment (jugmac00) [inline CHANGES.md:12]: `istead if` -> `instead of`

File: src/batou/lib/python.py [hunk 2]
```diff
@@ -226,3 +229,53 @@ def update(self):
...
+            self.requirements +
+            self.python_version +
+            open(__file__, 'r').read())
+        hash_content = hash_content.encode('utf-8')
+        self.env_hash = hashlib.new('sha256', hash_content).hexdigest()
+        self.env_dir = os.path.join('.appenv', self.env_hash)
+        self.env_ready = os.path.join('.appenv', self.env_hash, 'appenv.ready')
+
+    def verify(self):
+        assert os.path.exists(self.env_ready)
```
Comment (zagy) [inline src/batou/lib/python.py:252]: While i like it spelled this way, I don't think it's a good idea. Asserts may be optimised away, see https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement: "The current code generator emits no code for an assert statement when optimization is requested at compile time." (e.g. -O)

Comment (wosc) [reply src/batou/lib/python.py:252]: I, too, remember reading about this the Python manual, ages ago. But is this really a realistic concern (as opposed to an idealistic one)? (I can see the headlines already: "batou does not work properly on PyPy"... ;)

The alternative `def assert_(condition): if not condition: raise AssertionError()` just does not have the same nice ring to it... we could play with synonyms though; `expect()` maybe?

Comment (ctheune) [reply src/batou/lib/python.py:252]: Well, I'd rather check that we run in a Python without -O. Pytest does something specific smart and -O is really not useful at all, it _only_ removes assertions and that isn't something that's actually done. batou as an application can make the assertion (heh) that it doesn't run in an optimizing interpreter.

File: CHANGES.md (hunk 1 above)
Comment (ctheune) [reply CHANGES.md:12]: Right. :)

**Comment by ctheune** [comment]
@wosc: right. it is definitely intended to replace both VirtualEnv/Package and buildout.

I did go through the list of things with @sweh and a colleague yesterday. Poetry seemed too ambitious. Pipfile is something we'd be interested in leveraging, but it isn't there yet.

Some tooling for 'write me a lockfile' will be helpful - but actually it's not that hard. I'll likely provide a sample shell script, but I don't see a good way to provide it as an embedded function from within batou at the moment ...

**Pull request merged**